### PR TITLE
pocl: init at 6.0

### PR DESCRIPTION
**Pull request by jansol**
Dropping `pkgs.pocl` into `hardware.graphics.extraPackages` makes my (znver3) CPU it show in `clinfo` and device properties etc are listed correctly. Hashcat works if PoCL is the only OpenCL platform it can see, e.g. because of `OCL_ICD_VENDORS`, or because there are no others installed.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by lap1nou** [comment]
Hello and thank you for your work @jansol,

Any eta when this pkgs will be merged ? It seems needed for `hashcat` to work with my Intel CPU, I tested your package and I got the same error as you:
```
[2024-02-25 15:42:54.374418037]POCL: in fn POclCreateCommandQueue at line 104:
  |   GENERAL |  Created Command Queue 6 (0xa030a0) on device 0
[2024-02-25 15:42:54.374447544]POCL: in fn POclCreateContext at line 233:
  |   GENERAL |  Created Context 5 (0xa02f30)
[2024-02-25 15:42:54.374458980]POCL: in fn POclCreateCommandQueue at line 104:
  |   GENERAL |  Created Command Queue 7 (0xa04fb0) on device 0
[2024-02-25 15:42:55.415755454]POCL: in fn llvm_codegen at line 215:
  |   GENERAL |  Linking final module
/nix/store/2ab5740x0cy1d74qvbpl5s28qikmppl5-binutils-2.40/bin/ld: cannot find -lgcc: No such file or directory
/nix/store/2ab5740x0cy1d74qvbpl5s28qikmppl5-binutils-2.40/bin/ld: cannot find -lgcc_s: No such file or directory
error: linker command failed with exit code 1 (use -v to see invocation)
Final linking of kernel gpu_decompress failed.
```

Have you found a workaround or something maybe ?

Regards.

**Comment by jansol** [comment]
@lap1nou Nope, haven't really had time to look into the NixOS-specifics of this.

**Comment by xddxdd** [comment]
I tried injecting library search path of `libgcc.a` and `libgcc_s.so` to `ld` command which fixed the problem. Hashcat works for me now.

Here is my updated package: https://gist.github.com/xddxdd/3627283c38e3db9a5e7f18d25d3a0bcb

**Comment by xddxdd** [comment]
Also pocl 5.0 has been released with support for LLVM 17.

**Comment by JohnRTitor** [comment]
@xddxdd looks good! Could you adopt this PR, since it looks like it's stale for a while?

**Comment by jansol** [comment]
Sorry for the long silence, finally had a moment to poke at this again. Now the package is up to date with the latest upstream release and both clinfo and hashcat seem to be happy with it OOTB.

@xddxdd I've incorporated the changes from your gist and credited you with the same author line you've been using in your nur repo -- let me know if you'd prefer something different.

I've also enabled building the experimental remote backend, since it does not require additional dependencies (except optionally rdma-core, but that is an exercise for another time) and it only gets loaded if explicitly configured with environment variables at runtime. In other words, unless you know how to configure the remote backend and manually do that, you should never notice its existence.

The Vulkan and CUDA backends would be interesting to enable in the future, but those probably should get package level flags since they are bound to increase the closure size by a fair amount.

**Comment by K900** [comment]
Vulkan definitely shouldn't, FWIW. vulkan-loader is tiny and self-contained.

**Comment by jansol** [comment]
> Vulkan definitely shouldn't, FWIW. vulkan-loader is tiny and self-contained.

Yeah, I'm more concerned about the `clspv` dependency. Which reminds me, PoCL now supports SPIR-V kernels if spirv-tools and llvm-spirv-translator are available, lemme add those in real quick...

**Comment by jansol** [comment]
Note in case someone has been testing this with hashcat: you'll want to delete both PoCL's **and** hashcat's kernel caches between builds: `~/.cache/pocl/kcache` and `~/.cache/hashcat/kernels`. Otherwise you'll get an error from PoCL complaining about an LLVM version mismatch and a segfault from hashcat.

**Review by xddxdd — APPROVED** [review]
LGTM

**Review by xddxdd — APPROVED** [review]

File: pkgs/by-name/po/pocl/package.nix [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  meta = with lib; {
+    description = "A portable open source (MIT-licensed) implementation of the OpenCL standard";
+    homepage = "http://portablecl.org";
+    license = licenses.mit;
+    maintainers = with maintainers; [
+      jansol
+      xddxdd
+    ];
+    platforms = platforms.linux ++ platforms.darwin;
+  };
```
Comment (JohnRTitor) [inline pkgs/by-name/po/pocl/package.nix:88]:
```suggestion
  meta = {
    description = "A portable open source (MIT-licensed) implementation of the OpenCL standard";
    homepage = "http://portablecl.org";
    changelog = "https://github.com/pocl/pocl/releases/tag/v${finalAttrs.version}";
    license = lib.licenses.mit;
    maintainers = with lib.maintainers; [
      jansol
      xddxdd
    ];
    platforms = lib.platforms.unix;
  };
```

File: pkgs/by-name/po/pocl/package.nix [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  '';
+in stdenv.mkDerivation (finalAttrs: {
+  pname = "pocl";
+  version = "6.0";
+
+  src = fetchFromGitHub {
+    owner = "pocl";
+    repo = "pocl";
+    rev = "v${finalAttrs.version}";
+    sha256 = "sha256-NHR9yrI6Odb/s/OBnEVifdcLEXvwqzAMZWpGViv2cJg=";
```
Comment (JohnRTitor) [inline pkgs/by-name/po/pocl/package.nix:43]:
```suggestion
    hash = "sha256-NHR9yrI6Odb/s/OBnEVifdcLEXvwqzAMZWpGViv2cJg=";
```

**Review by JohnRTitor — APPROVED** [review]
Minor nitpicks before I merge this :) Additionally the whole thing can be formatted with `nixfmt-rfc-style`.

**Comment by JohnRTitor** [comment]
Result of `nixpkgs-review pr 261736` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>pocl</li>
  </ul>
</details>

**Comment by jansol** [comment]
@JohnRTitor Thanks for taking care of those!

**Pull request merged**